### DISCUSSION TITLE
[Bug]: decodeScan - unexpected MCU data error for scanned pdf documents

### DISCUSSION THREAD

**Discussion by DongusJr** [mozilla/pdf.js#18835] [Q&A]
### Attach (recommended) or Link to PDF file
[scanned-document-error.pdf](https://github.com/user-attachments/files/17230340/scanned-document-error.pdf)

### Web browser and its version
Chrome
Version 129.0.6668.58 (Official Build) (arm64)

### Operating system and its version
MacOS - Sonoma 14.6.1

### Is the bug present in the latest PDF.js version?
Yes

### Is a browser extension
No

### Steps to reproduce the problem
with https://mozilla.github.io/pdf.js/web/viewer.html open developer tools in your browser of choice (I used Chrome).
Open the attached document and notice that in the developer tools console we get the error
`Warning: decodeScan - unexpected MCU data, current marker is: ffff` around 600~ times

### What is the expected behavior?
Not getting warning logs for decodeScan

### What went wrong?
In the viewer.html, this does not cause any noticeable disruption to the document.
However, in a React application that uses pdf.js (in my case it was version 4.5.136 with legacy), this error sometimes causes extreme load times and in worst case scenarios, completely blank pdf or crashes in the browser.
This seems to happen more with larger documents.

I'll leave a React pdf component that reproduces this error (This code is just to give an idea how the pdf is rendered in our application):
```tsx
import {
  getDocument,
  GlobalWorkerOptions,
} from 'pdfjs-dist/legacy/build/pdf.mjs';
import pdfWorker from 'pdfjs-dist/legacy/build/pdf.worker.mjs?url';
import React, { useEffect, useRef } from 'react';

type SimplePdfViewerProps = {
  contentBase64: string;
};

export const SimplePdfViewer = ({ contentBase64 }: SimplePdfViewerProps) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  const base64ToByteArray = (byteData: string) => {
    const bytes = window.atob(byteData);
    const byteLength = bytes.length;
    const byteArray = new Uint8Array(new ArrayBuffer(byteLength));
    for (let i = 0; i < byteLength; i++) {
      byteArray[i] = bytes.charCodeAt(i);
    }
    return byteArray;
  };

  useEffect(() => {
    if (typeof window !== 'undefined' && 'Worker' in window) {
      GlobalWorkerOptions.workerSrc = new URL(
        pdfWorker,
        import.meta.url,
      ).toString();
    }

    const loadDocument = async (arr: Uint8Array) => {
      const pdf = await getDocument({
        data: arr,
      }).promise;

      const page = await pdf.getPage(1);

      const canvas = canvasRef.current;

      if (!canvas) return;

      const context = canvas.getContext('2d');

      if (context === null) return;

      const canvasViewport = page.getViewport({ scale: 1 });
      canvas.height = canvasViewport.height;
      canvas.width = canvasViewport.width;

      await page.render({
        canvasContext: context,
        viewport: canvasViewport,
      }).promise;
    };

    const typedArr = contentBase64 && base64ToByteArray(contentBase64);
    if (typedArr) {
      loadDocument(typedArr);
    }
  }, []);

  return (
    <canvas
      ref={canvasRef}
      style={{
        width: '100%',
      }}
    ></canvas>
  );
};

```

### Additional context

I have encountered this issue with few documents, I am not sure if the problem lies in the pdf document it self or if this is a bug in pdf.js.

I have tried validating the pdf to check if it is corrupt but everything seems fine.

The warning is thrown in src/core/jpg.js, which might be because the pdf document is scanned.

Any idea what precisely causes this?

**Comment by calixteman** [comment]
You can change the verbosity level:
https://github.com/mozilla/pdf.js/blob/567df4214983a12da9f3d02a2ab1ea646f0dfcf1/web/app_options.js#L429-L433

Do you have a pdf you can share with us causing a crash or blank page ?
If you can't make it public you can send it to me on my pro email (I'm the triage owner of the pdf component: https://bugzilla.mozilla.org/enter_bug.cgi?product=Firefox&component=PDF%20Viewer).

**Reply by DongusJr** [reply]
I have a larger file here:
[crash-error-document-larger.pdf](https://github.com/user-attachments/files/17233243/crash-error-document-larger.pdf)

However, when testing this file I noticed that it was not crashing the browser in the test application that I made to reproduce this issue. However, it crashes the original solution that originally lead me to investigating this problem.
I will have to debug this, Ill add a comment if I figure out why it only crashes in the original application.

one more thing,
as I am not knowledgable about the inner workings of rendering pdfs/jpgs. What generally causes `decodeScan - unexpected MCU data` error? I have noticed that this warning always appears on documents that crash/don't render
